cassandra-five-zero-types/src/main/java/org/apache/cassandra/spark/reader/ComplexTypeBuffer.java

```diff
@@ -0,0 +1,118 @@
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.cassandra.db.marshal.AbstractType;
+import org.apache.cassandra.db.marshal.ByteBufferAccessor;
+import org.apache.cassandra.db.marshal.ListType;
+import org.apache.cassandra.db.marshal.MapType;
+import org.apache.cassandra.db.marshal.SetType;
+import org.apache.cassandra.db.marshal.UserType;
+import org.apache.cassandra.db.rows.Cell;
+import org.apache.cassandra.serializers.CollectionSerializer;
+
+/**
+ * ComplexTypeBuffer is a util class for reconstructing multi-cell data into complex types such as unfrozen lists, maps, sets, or UDTs.
+ * ComplexTypeBuffer buffers all the cell ByteBuffers then reconstructs as a single ByteBuffer.
+ */
+public abstract class ComplexTypeBuffer
+{
+    private final List<ByteBuffer> buffers;
+    private final int cellCount;
+    private int length = 0;
+
+    public ComplexTypeBuffer(int cellCount, int bufferSize)
+    {
+        this.cellCount = cellCount;
+        this.buffers = new ArrayList<>(bufferSize);
+    }
+
+    public static ComplexTypeBuffer newBuffer(AbstractType<?> type, int cellCount)
+    {
+        ComplexTypeBuffer buffer;
+        if (type instanceof SetType)
+        {
+            buffer = new SetBuffer(cellCount);
+        }
+        else if (type instanceof ListType)
+        {
+            buffer = new ListBuffer(cellCount);
+        }
+        else if (type instanceof MapType)
+        {
+            buffer = new MapBuffer(cellCount);
+        }
+        else if (type instanceof UserType)
+        {
+            buffer = new UdtBuffer(cellCount);
+        }
+        else
+        {
+            throw new IllegalStateException("Unexpected type deserializing CQL Collection: " + type);
+        }
+        return buffer;
+    }
+
+    public void addCell(Cell cell)
+    {
+        add(cell.buffer());  // Copy over value
+    }
+
+    void add(ByteBuffer buffer)
+    {
+        buffers.add(buffer);
+        length += buffer.remaining();
+    }
+
+    ByteBuffer build()
+    {
+        ByteBuffer result = ByteBuffer.allocate(Integer.BYTES + (buffers.size() * Integer.BYTES) + length);
+        result.putInt(cellCount);
+        for (ByteBuffer buffer : buffers)
+        {
+            result.putInt(buffer.remaining());
+            result.put(buffer);
+        }
+        // Cast to ByteBuffer required when compiling with Java 8
+        return (ByteBuffer) result.flip();
+    }
+
+    /**
+     * Pack the cell ByteBuffers into a single ByteBuffer using Cassandra's packing algorithm.
+     * It is similar to {@link #build()}, but encoding the data differently.
+     *
+     * @return a single ByteBuffer with all cell ByteBuffers encoded.
+     */
+    public ByteBuffer pack()
+    {
+        // See CollectionSerializer.deserialize for why using the protocol v3 variant is the right thing to do.
+        // TODO(c4c5): Do we need to specify ProtocolVersion.V3 as last parameter?
```

Comment:
The ProtocolVersion can be removed safely. See https://issues.apache.org/jira/browse/CASSANDRA-18114
Can you remove the comment at line#109 (that comes from the four-zero subproject)? To avoid any confusion.